File: src/api/recolor.js
import { JSDOM } from "jsdom";
import XMLSerializer from "xmlserializer";

import { hexToHsl, hslToHex } from "./convert.js";
import { adjustColor, updateStops } from "./colorManagement.js";
import createOverlayGradient from "./createOverlayGradient.js";
import addTextElement from "./addTextElement.js";

const baseStops = [
    "#FABA1B","#F0B122", // shadow
    "#FABA1B","#F0B122", // back
    "#FFEDBA","#FFDC78", // shine
    "#FFE69C","#FFCA38", // front
];

const baseHSL = baseStops.map(hexToHsl);
const ref = baseHSL[2];
const offsets = baseHSL.map(([h,s,l]) => ({ dh: h - ref[0], ds: s - ref[1], dl: l - ref[2] }));

function recolor(
    data,
    style = "shaded",
    varient = "left1",
    baseColor = "#ffca38",
    backColor = "#000000",
    iconColor = "#000000",
    mediumIcon = "",
    smallIcon = "",
    text = "",
    saturation = parseFloat("100") / 100,
    brightness = parseFloat("100") / 100,
    contrast = parseFloat("100") / 100,
    isCustomBackColor = false,
    isCustomIconColor = false
) {
    let [h, s, l] = hexToHsl(baseColor);

    s = Math.min(1, Math.max(0, s * saturation));
    l = Math.min(1, Math.max(0, l * brightness));
    l = 0.5 + (l - 0.5) * contrast;

    const newColors = offsets.map(o =>
        hslToHex(
            (h + o.dh + 1) % 1,
            Math.min(1, Math.max(0, s + o.ds)),
            Math.min(1, Math.max(0, l + o.dl))
        )
    );

    const dom = new JSDOM(data, { contentType: "image/svg+xml" });
    const svgDoc = dom.window.document;

    updateStops(svgDoc, newColors, isCustomBackColor, backColor);

    const overlayId = "overlayGradient";
    const baseOverlayColor = isCustomIconColor ? iconColor : baseColor;
    const adjustedOverlayColor = adjustColor(baseOverlayColor, saturation, brightness, contrast);
    const overlayFill = isCustomIconColor
        ? baseOverlayColor
        : createOverlayGradient(svgDoc, overlayId, adjustedOverlayColor);

    let mediumIconY = "69%";

    if (style === "shaded" && varient === "center1") {
        mediumIconY = "72%"
    }

    addTextElement(svgDoc, "50%", mediumIconY, "112", mediumIcon, overlayFill);
    addTextElement(svgDoc, "87%", "73%", "96", smallIcon, overlayFill, "end");
    addTextElement(svgDoc, "88%", "75%", "68", text, overlayFill, "end");

    return XMLSerializer.serializeToString(svgDoc.documentElement);
}

export default recolor;

File: src/cornelia/scripts/folders/recolor.py
import xml.etree.ElementTree as ET

from src.cornelia.scripts.folders.color_management import (
    adjust_color,
    update_stops,
    update_color,
    darken_color,
    hex_to_hsl,
    hsl_to_hex
)
from src.cornelia.scripts.folders.create_overlay_gradient import create_overlay_gradient
from src.cornelia.scripts.folders.add_text_element import add_text_element

SVG_NS = "http://www.w3.org/2000/svg"
ET.register_namespace("", SVG_NS)

base_stops = [
    "#FAC63E", "#F2C146",  # shadow
    "#FAC63E", "#F2C146",  # back
    "#FFF7DE", "#FFE79E",  # shine
    "#FFF0C2", "#FFD65C",  # front
]

base_hsl = [hex_to_hsl(c) for c in base_stops]
ref = base_hsl[-1]

offsets = [
    {
        "dh": h - ref[0],
        "ds": s - ref[1],
        "dl": l - ref[2],
    }
    for (h, s, l) in base_hsl
]

def recolor(
    data: str,
    style="shaded",
    variant="left1",
    base_color="#FFD65C",
    back_color="#000000",
    icon_color="#000000",
    medium_icon="",
    small_icon="",
    text="",
    saturation=1,
    brightness=1,
    contrast=1,
    is_custom_back_color=False,
    is_custom_icon_color=False,
):
    root = ET.fromstring(data.encode("utf-8"))
    h, s, l = hex_to_hsl(base_color)

    s = max(0, min(1, s * saturation))
    l = max(0, min(1, l * brightness))
    l = 0.5 + (l - 0.5) * contrast

    overlay_fill = None

    if style == "shaded":
        new_colors = [
            hsl_to_hex(
                (h + o["dh"]) % 1,
                max(0, min(1, s + o["ds"])),
                max(0, min(1, l + o["dl"]))
            )
            for o in offsets
        ]

        adjusted_back_color = adjust_color(
            back_color, saturation, brightness, contrast
        )

        update_stops(root, new_colors, is_custom_back_color, adjusted_back_color)

        base_overlay_color = icon_color if is_custom_icon_color else base_color
        adjusted_overlay_color = adjust_color(
            base_overlay_color, saturation, brightness, contrast
        )

        overlay_fill = (
            adjusted_overlay_color
            if is_custom_icon_color
            else create_overlay_gradient(root, "overlayGradient", adjusted_overlay_color)
        )

    else:
        adjusted_base = adjust_color(base_color, saturation, brightness, contrast)

        adjusted_back = (
            adjust_color(back_color, saturation, brightness, contrast)
            if is_custom_back_color
            else adjust_color(darken_color(adjusted_base, 0.62),
                              saturation, brightness, contrast)
        )

        update_color(root, adjusted_base, adjusted_back)

        overlay_fill = (
            adjust_color(icon_color, saturation, brightness, contrast)
            if is_custom_icon_color
            else adjusted_back
        )

    """
    ————————————————————————————————————————————————————————————————
    Icon/text position
    ———————————————————————————————————————————————————————————————— 
    """

    medium_icon_y = "69%"
    if variant == "center1":
        medium_icon_y = "72%"

    add_text_element(root, "50%", medium_icon_y, "112", medium_icon, overlay_fill)
    add_text_element(root, "87%", "73%", "96", small_icon, overlay_fill, "end")
    add_text_element(root, "88%", "75%", "68", text, overlay_fill, "end")

    ET.register_namespace("", SVG_NS)
    return ET.tostring(root, encoding="unicode", method="xml")